sips/sip-018/sip-018-signed-structured-data.md

```diff
@@ -0,0 +1,204 @@
+
+## Formal specification
+
+To obtain a signature *S* of structured data *D* using private key *K*:
+
+- `S = signStructuredData(D, K)`; where,
+
+- `signStructuredData(D, K) = sign(messageHash(D), K)`; where `sign` produces a `secp256k1` signature of a message hash using private key *K*; and,
+
+- `messageHash(D) = sha256(structuredDataPrefix || domainHash || structuredDataHash(D))`; where `structuredDataPrefix` is a static value of `"\xC0"`, and `domainHash` and `structuredDataHash` are as described below.
+
+**Definition of** `domainHash`
+
+The domain hash ensures that messages are chain and application specific. A `domainHash` is generated out of a `domain` Clarity Value:
+
+- `domainHash = structuredDataHash(domain)` where `structuredDataHash` and `domain` are as described below.
+
+**Definition of** `structuredDataHash`
+
+The `structuredDataHash` function calculates a hash *H* of the input Clarity Value *D*.
+
+- `H = structuredDataHash(D)`; where,
+- `structuredDataHash(D) = sha256(ToCVWireFormat(D))`; where `ToCVWireFormat` is a function that encodes an input Clarity Value to its wire format representation as described in *SIP005*.
+
+**Definition of** `domain`
+
+The `domain` is a `tuple` type Clarity Value that represents the domain in which messages are signed. Its Clarity type definition is as follows:
+
+```clojure
+{
+	name: (string-ascii len)
+	version: (string-ascii len)
+	chain-id: uint
+}
+```
+
+`name` contains a readable application name for which Structured Data is being signed.
+
+`version` contains a string representation of the application version. The application developer may choose to change this value to ensure that signed Structured Data of one version of the application is incompatible with another version.
+
+`chain-id` contains the ID of the chain the Signed Structured Data is targeting.
+
+The length `len` of the string fields are not important as long as the resulting `domain` Clarity Value does not exceed the permitted maximum length. It is encouraged that wallet applications display this information to the user. Other fields may be added per the discretion of the application developer. For example, another differentiating field may be a `verifying-contract` that contains the principal of the contract performing the signature validation.
+
+**Definition of Structured Data**
+
+Structured Data can be any valid Clarity Value except `buff`. It is assumed that it will usually take the form of a `tuple` type, as it allows for multiple fields with readable names to be encoded. The standard does permit it to be another type to remain flexible. Still, developers should consider the informational content of requesting a user to sign a simple type like a `principal`, `int`, `uint`, or `bool`. A tuple can of course contain any type, including `buff`.
+
+A future SIP may define standard types of Structured Data.
+
+## Collisions and replay attacks
+
```

Comment:
So, notwithstanding the recommendation that discussions pertaining to how to safely encode Clarity data belongs in a separate SIP, it's not the case that the transaction version is limited to `0x00` or `0x80` in the future.  In fact, the transaction codec only requires that the leading bit be set in the transaction in order for it to be considered a testnet transaction.  So, `0x01` is a valid mainnet transaction version, as are `0x02..0x7f`.  Bytes `0x80...0xff` are for testnet.

Reply:
If you want a prefix that is guaranteed to never be a valid transaction, you could use the byte sequence `0x00 0x00 0x00 0x00 0x00 0x00`.  This encodes a mainnet version (first byte), a chain ID of 0 (next 4 bytes), and a currently-invalid transaction authorization type (`0x00` -- only `0x04` and `0x05` are supported, per SIP 005).

Reply:
Would you recommend such a prefix instead of a single byte? Is there a non-zero prefix we could use and specifically reserve it for signed messages?

Reply:
Sure, this SIP could propose reserving a 1-byte prefix, and declare that this value can never be a transaction version.  How about `0xff`, or `0x7f`?

Reply:
To be clear, we'll need to add this to Stacks 2.1 so it will never accept a transaction with this version (but now is the time to do that).  If you go this route, please also update the Activation section to require that Stacks 2.1 ships with this change.

Reply:
I'm partial to `0x7f` if there is a strong reason we shouldn't the already used `0xc0`. Also, I feel this SIP could activate without requiring Stacks 2.1 block the prefix byte as it is currently unused. As a comparison, EIP712's prefix of `[0x19, 0x01]` was also not explicitly blocked in the node as far as I know. It was chosen because no valid RLP encoding produces a byte string starting with `0x19`.

Reply:
> Also, I feel this SIP could activate without requiring Stacks 2.1 block the prefix byte as it is currently unused. 

We don't have a choice.  The network will accept a transaction with version byte `0x7f` today.  If we're going to reserve a byte to indicate a SIP-018 payload, then we must also update Stacks so it will _never_ accept a transaction that starts with that byte.  Otherwise you run the risk of malicious SIP-018 apps tricking users into signing well-formed transactions.

Reply:
Ah, I didn't know that! I thought it would reject the transaction if the first/version byte were `0xc0` (or `0x7f` for that matter). I should've checked that instead of making the assumption. So the safest prefix to pick right now would be something like `[0xc0, 0x00, 0x00, 0x00, 0x00, 0xc0]` or `[0x7f, 0x00, 0x00, 0x00, 0x00, 0x7f]`.

Given that full support cannot happen before Stacks 2.1 any (due to the inability to hash principals), we could try for the solution you suggested before; that is, block that specific version byte in the node. What do you think? I'm still leaning towards a single prefix byte over six. Seems more optimal. But whatever we go for, in the end it is just a matter of updating the SIP and reference implementations.

Reply:
I think one of these options would work:

* Stipulate a 6-byte prefix with the above constraints 
* Stipulate a 6-byte prefix, and use the first 5 bytes to encode something useful a versioning scheme, a payload length, a vanity string, or what-not.
* Stipulate a 1-byte prefix that activates with 2.1, *and* add a very explicit warning to this SIP that implementors _should not_ eagerly implement this SIP until Stacks 2.1 (given the security risk)

Reply:
The last one is not great because SIP018 can definitely be used today. To add, Hiro is actively implementing the standard right now so  we are going to see these messages very soon. Once that happens it might be painful to change the prefix.
- https://github.com/hirosystems/connect/issues/225
- https://github.com/hirosystems/stacks-wallet-web/issues/2460

>- Stipulate a 6-byte prefix, and use the first 5 bytes to encode something useful a versioning scheme, a payload length, a vanity string, or what-not.

I considered this as well but the domain already contains that information. Domain objects are easier to assemble for end devs over remembering the right encoding for the prefix. A static prefix is probably a lot easier.
```clarity
(to-consensus-buff {domain: "My App", chain-id: chain-id, version: "1.0.0"})
```

How about this prefix: `[0x53, 0x49, 0x50, 0x30, 0x31, 0x38]`? It spells `SIP018`.

Reply:
To add: the readable prefix should make the reference implementation and those of end devs straightforward, e.g.:

```clarity
(define-constant sip018-prefix (to-buff "SIP018"))
(define-constant domain-hash (sha256
 {
  name: "My App",
  chain-id: chain-id,
  version: "1.0.0"
 }
))
(define-constant message-hash-prefix (concat sip018-prefix domain-hash))

(define-read-only (sip018-message-hash (payload {some-type}))
  (sha256 (concat message-hash-prefix (sha256 (to-consensus-buff payload))))
)
```

Reply:
> The last one is not great because SIP018 can definitely be used today. To add, Hiro is actively implementing the standard right now so we are going to see these messages very soon.

Yeah, that's what I thought.  Looks like it needs to be option 1 or 2.

> How about this prefix: [0x53, 0x49, 0x50, 0x30, 0x31, 0x38]? It spells SIP018

Sounds good :+1: